lib/msf/core/payload/adapter/fetch/http.rb

```diff
@@ -10,7 +10,11 @@ def initialize(*args)
   end
 
   def cleanup_handler
-    cleanup_http_fetch_service(@fetch_service, @delete_resource)
+    if @fetch_service
```

Comment:
This change in isolation looks good to me; But this `cleanup_http_fetch_service` implementation that gets called looks very odd to me:

https://github.com/rapid7/metasploit-framework/blob/50579fb1de8fad4796a2a0ccd4001590efad2018/lib/msf/core/payload/adapter/fetch/server/https.rb#L44-L61

I don't think there should be two calls to `deref` in a single cleanup call, is there another issue hidden here maybe? 🤔

Reply:
Yeah the ref counting is wrong here; Verified with some extra logging to rex-core:

```diff
diff --git a/lib/rex/sync/ref.rb b/lib/rex/sync/ref.rb
index 50b0146..9087747 100644
--- a/lib/rex/sync/ref.rb
+++ b/lib/rex/sync/ref.rb
@@ -45,6 +45,8 @@ module Ref
     @_references       = 1
     @_references_mutex = Mutex.new
 
+    $stderr.puts "[ref-count] calling refinit: #{caller}, total references: current #{@_references}"
+
     self
   end
 
@@ -54,6 +56,7 @@ module Ref
   def ref
     @_references_mutex.synchronize {
       @_references += 1
+      $stderr.puts "[ref-count] creating ref: #{caller}, total references: current #{@_references}"
     }
 
     self
@@ -65,11 +68,15 @@ module Ref
   #
   def deref
     @_references_mutex.synchronize {
-      if ((@_references -= 1) == 0)
+      @_references -= 1
+      $stderr.puts "[ref-count] creating deref: #{caller}, total references: #{@_references}"
+      if (@_references == 0)
+        $stderr.puts "[ref-count] calling cleanup"
         cleanup
 
         true
       else
+        $stderr.puts "[ref-count] no cleanup required"
         false
       end
     }
```

We end up with a refcount of `-1` as far as I can see from a quick glance

Reply:
Yeah that looks like an issue, but I copied it over from @bwatters-r7's original implementation [here](https://github.com/rapid7/metasploit-framework/blob/08beaddf0aa9fa5994e9e31350686c38f3a4e103/lib/msf/core/payload/adapter/fetch/server/https.rb#L44). I tested it again by removing everything after the `fetch_service.deref` line and confirmed I can start two services then tear them down individually and the HTTP server will remain until the last handler job is killed which is the expected behavior. Based on that, I don't think the explicit call to `#cleanup` is necessary, it looks like `#deref` handles it automatically.

Reply:
I always thought this was wonky; I never understood why we needed to `deref` until empty, the `deref` again.  I know in theory, it should call its own cleanup when the refcount is 0, but that expectation did not match reality.
I will go back and take a look, but I _know_ that I could not start a second service without the second deref, and I am pretty sure that the service was never stopped.  I spent a long time trying to figure out what was going on, and cursed the rex-sync library repeatedly.

Reply:
@bwatters-r7 are we alright to keep the proposed changes or are you still looking into why it needed to be done this way?

Reply:
I started looking at the old commits and was trying to remember what tests I ran that made me think this was an issue.  It is on my list for today to see if I can figure it out.

Reply:
I found the conversation where I think I was working on this.  I was unable to recreate the issue when I removed
```
     if fetch_service.resources.empty? 
       # if we don't call deref, we cannot start another httpserver 
       # this is a reimplementation of the cleanup_service method 
       # in Exploit::Remote::SocketServer 
       temp_service = fetch_service 
       fetch_service = nil 
       temp_service.cleanup 
       temp_service.deref 
     end 

```
That said, I also can not longer use the `_servicemanager` command....  it was hidden behind a feature flag that does not exist, and I cannot find a built-in command that matches..... It is also late, so I might be misremembering or completely missing it.  I'll look again tomorrow, but this looks like an overcorrection on my part right now.